File: app/models/topic.rb
# typed: strict
# frozen_string_literal: true

class Topic < T::Struct
  extend T::Sig

  include T::Struct::ActsAsComparable

  # 名前の最大文字数 (値に強い理由は無い)
  NAME_MAX_LENGTH = 30
  # 説明の最大文字数 (値に強い理由は無い)
  DESCRIPTION_MAX_LENGTH = 150

  const :database_id, Types::DatabaseId
  const :number, Integer
  const :name, String
  const :description, String
  const :visibility, TopicVisibility
  const :can_update, T.nilable(T::Boolean)
  const :can_create_page, T.nilable(T::Boolean)
  const :space, Space

  sig { returns(T::Boolean) }
  def visibility_public?
    visibility == TopicVisibility::Public
  end

  sig { returns(T::Boolean) }
  def can_update?
    can_update.not_nil!
  end

  sig { returns(T::Boolean) }
  def can_create_page?
    can_create_page.not_nil!
  end
end

File: go/internal/model/topic.go
package model

import (
	"time"
)

// TopicVisibilityはトピックの公開範囲を表す
type TopicVisibility int32

const (
	// TopicVisibilityPublicは公開トピック
	TopicVisibilityPublic TopicVisibility = 0
	// TopicVisibilityPrivateは非公開トピック
	TopicVisibilityPrivate TopicVisibility = 1
)

// Topicはトピックのドメインモデル
type Topic struct {
	ID          TopicID
	Space       *Space
	Number      int32
	Name        string
	Description string
	Visibility  TopicVisibility
	DiscardedAt *time.Time
}

// トピックのフォームが送信する公開範囲の値。
const (
	topicVisibilityPublicValue  = "public"
	topicVisibilityPrivateValue = "private"
)

// ParseTopicVisibilityはトピックのフォームが送信する値をTopicVisibilityに変換する。
// 2つ目の戻り値はその値がフォームの提示する2つのいずれかであるかを返す。それ以外の値を、
// 送信者が選んでいない公開範囲へ倒さずに拒否できるようにするためである。
func ParseTopicVisibility(value string) (TopicVisibility, bool) {
	switch value {
	case topicVisibilityPublicValue:
		return TopicVisibilityPublic, true
	case topicVisibilityPrivateValue:
		return TopicVisibilityPrivate, true
	default:
		return TopicVisibilityPublic, false
	}
}

// Stringはトピックのフォームが公開範囲として送信する値を返す。
func (v TopicVisibility) String() string {
	if v == TopicVisibilityPrivate {
		return topicVisibilityPrivateValue
	}
	return topicVisibilityPublicValue
}
